### PR TITLE
[FEATURE] Add MigrateBackendModuleRegistrationRector

### PR DESCRIPTION
**Pull request by simonschaufi**
Resolves: #3347

### CODE REVIEW

**Comment by simonschaufi** [comment]
TODO:
- Read existing array
- Maybe parse module call into array variables?
- Render array and write it into the Modules.php file

**Comment by simonschaufi** [comment]
@samsonasik Dear Abdul, this is probably the most difficult Rector rule ever you have seen so far because I not only need to change code within a file but across different files. Creating files and writing to it is no problem for me as we have several other rules where we do this already. The big challenge here is to parse an existing php array and add a new item into it.

My initial thought was that i just create the new php array in pure php and then let the php parser do its thing and then I just merge the 2 nodes together but this isn't working because of some internal things I don't understand. For some reason the old nodes get duplicated even though the nodes are properly merged when you debug the node tree. My guess is that is has to do with the attributes that get attached to each node but I'm not sure.

My second try was to create the ArrayItem by hand but then I have the issue that the written output is just a one liner (see the 2 versions below the comment `Merge the arrays`).

Maybe you have some deeper understanding on what is going on internally in the php parser and have an idea that could solve my issue here?

File: rules/TYPO312/v0/MigrateBackendModuleRegistrationRector.php [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+        /** @var Array_ $array */
+        $array = $return->expr;
+
+        return $array;
+    }
+
+    private function printFile(File $file, string $filePath) : void
+    {
+        // only save to string first, no need to print to file when not needed
+        $newContent = $this->formatPreservingPrinter->printParsedStmstAndTokensToString($file);
```
Comment (samsonasik) [inline rules/TYPO312/v0/MigrateBackendModuleRegistrationRector.php:493]: this removed at 

- https://github.com/rectorphp/rector-src/pull/6036

use directly from `betterStandardPrinter` instead:

```php
        $newContent = $this->betterStandardPrinter->printFormatPreserving(
             $file->getNewStmts(),
            $file->getOldStmts(),
            $file->getOldTokens()
       );
```

**Comment by simonschaufi** [comment]
@samsonasik thank you very much for looking into it bit even with this change the problem still exists that the output is currently still the duplicate of `tools_toolsmaintenance` instead of the added `tools_toolssettings`

**Comment by samsonasik** [comment]
the array structure may be appended, the structure may be need to be unset and append by index, like using `array_splice()`...

**Comment by simonschaufi** [comment]
I'm merging the two arrays here: https://github.com/sabbelasichon/typo3-rector/pull/4243/files#diff-6659bbad2dc4a8d574769de61529b6e6d9387889d39813bbe2f246c873ba3bfeR461

When you inspect the code afterwards, it looks correct but the generated code is not correct and that is strange. Is there somebody in the Rector team who has more understanding on what is going on under the hood?

**Comment by simonschaufi** [comment]
ping @samsonasik @TomasVotruba

**Pull request merged**